### ISSUE TITLE
vk.xml: VkShaderModuleCreateInfo's pCode member has some garbage in its "len" attribute

### ISSUE THREAD

**Issue by yzsolt** [KhronosGroup/Vulkan-Docs#555]
It looks like this in the latest vk.xml (line 846):
`<member len="latexmath:[codeSize \over 4]">`
Should be simply codeSize without the LaTeX garbage.

**Comment by krOoze** [comment]
The `codeSize` is in (single) bytes, but `pCode` is in `uint32_t` (4-bytes elements). So as it currently is defined it must be `codeSize/4`.
What exactly are you suggesting here?

**Comment by yzsolt** [comment]
I didn't know it was intended, my bad.
Still, it seems weird to see something like that in the XML. The other len values are just simply the names of the corresponding members. I use this info to generate nice span-parameter setters for struct members (something like Vulkan-Hpp does with ArrayProxies for array command parameters). However, in its current form this is unnecessarily hard to parse for a Vulkan wrapper code generator.
I think the relation between the array's length and the pointed type's size could be documented in an additional attribute instead, and this purely display-related latexmath code could be then generated based on that attribute by the documentation generator.

**Comment by krOoze** [comment]
> I didn't know it was intended, my bad.

Yes, I suspected that is the case. Wasn't sure though... One could argue `codeSize` could have been size in `uint32_t`s, or `pCode` could have been `uint8_t` array. But that cannot be changed before Vulkan 2.0.

> Still, it seems weird to see something like that in the XML.

If you do not like this, you are gonna love `VkPipelineMultisampleStateCreateInfo::pSampleMask` 😈 

If you are using `vk.xml` you shoud be reading [vk.xml schema spec](https://www.khronos.org/registry/vulkan/specs/1.0/registry.html). It is documented there `len` can be a latex expression. And documents other gotchas.

> I think the relation between the array's length and the pointed type's size could be documented in an additional attribute instead

How? `len` is already an/the additional attribute.

One thing that could be done is to move them to `noautovalidity`. I don't see the benefit though. You could equally just ignore `len` as soon as you see `latexmath:`.

**Comment by krOoze** [comment]
PS: It is not that hard to parse.
You could just hardcode those two latexmath cases. New ones should not appear frequently, and it would be a 5 min fix for your parser if they do.

**Comment by oddhack** [comment]
An alternate representation could be provided, but the alternative representation would *also* have to be parsed. It might be easier to parse

> ceil(rasterizationSamples/32)

I'm OK with having an altlen="some syntax" tag for external users, but I don't know what "some syntax" should be offhand. Maybe C with imbedded parameter names and arbitrary ISO math calls?ANSI -libm calls.

**Comment by yzsolt** [comment]
> If you are using vk.xml you shoud be reading vk.xml schema spec.
Yes, I should have done that and I certainly will. It was still useful to give it a go though.
> How? len is already an/the additional attribute.
Given the `VkPipelineMultisampleStateCreateInfo` example I should reconsider my initial idea. Maybe a simplified formula, like `codeSize/4` or `ceil(rasterizationSamples/32)` would be still useful though.
> You could just hardcode those two latexmath cases.
That's what I'd have liked to avoid as there are already some other things (mainly C declarations) which require non-trivial parsing, but I don't see a better way neither.

**Comment by krOoze** [comment]
@oddhack 
There is a bit of a confusion about it though:
1) The `vk.xml` uses `latexmath:` in the `len` attribute
2) The schema spec says for `member::len` it can be a latex expression delimited by `\($ $\)`
3) The schema spec says for `param::len` it can be "a simple expression" prefixed with `math:`
4) The [validitygenerator.py#L370](https://github.com/KhronosGroup/Vulkan-Docs/blob/bbb637d6dbe78a5d8c1c8f8f2129353b0e4dc1dc/src/spec/validitygenerator.py#L370) seems to expect `latex:` prefix.

**Comment by oddhack** [comment]
@krOoze (2) and (3) are out of date math markup, and will be fixed in the next public spec update. (4) is an issue for @TobiasHector to tweak in those scripts, although I don't think any current VUs are actually going down that path in the code - hard to be certain though.

**Comment by oddhack** [comment]
And I'll see about adding an altlen= attribute with a C notation as suggested above.

**Comment by krOoze** [comment]
@oddhack (4) is just trivial error I am sure. It concerns the comma separated lengths in `vk.xml`. It is not even used in reality yet.

**Comment by oddhack** [comment]
I believe @krOoze's issues on the XML / registry schema spec are addressed now. I'll keep this open to look at the altlen= attribute.

**Comment by oddhack** [comment]
In the 1.0.61 spec update, the XML schema supports, and provides for the two examples, an 'altlen' attribute intended to be more easily parseable for code generators.

**Comment by krOoze** [comment]
@oddhack I am not sure `altlen="ceil(rasterizationSamples / 32)"` is desirable. It does not say the same thing in C as in "real math". What this would actually mean in C is `double altlen = ceil( double(rasterizationSamples >> 5) )`, which is semantically different than what is in `len`.

**Comment by ratchetfreak** [comment]
really it means `floor((rasterizationSamples+31) / 32)` in C or `(int)ceil(rasterizationSamples / 32f)`

**Comment by krOoze** [comment]
Damn, I meant `rasterizationSamples / 64 + 1`. Corrected.

Nevermind,
the point is the `altlen` should (and probably was meant to be) a C that can be dumped in a C program worry free. `altlen="ceil(rasterizationSamples / 32)"` would not work.

Oh, and just noticed:
`altlen` seems to be missing for `param` in `registry.rnc`.

**Comment by oddhack** [comment]
@krOoze @ratchetfreak we're open to tweaking the altlen syntax in ways that make it easier for you to consume; please propose a PR.

**Comment by krOoze** [comment]
@oddhack Your wish is my command:  ✨ #566.
